File: brainfuck.py
#A simple (probably) Brainfuck interpreter.
#If does not work, blame Jason

import sys

memcells = [0]*30000
memptr = 0

def char_lt():
    """
    performs a <
    """
    global memptr
    if memptr <= 0:
        raise (ValueError, "Trying to point before starting of memory")
    memptr -= 1

def char_gt():
    """
    performs a >
    """
    global memptr
    if memptr >= 3*10**4 - 1:
        raise (ValueError, "Trying to point outside memory")
    memptr += 1

def char_plus():
    """
    performs a +
    """
    global memptr
    memcells[memptr] = (memcells[memptr] + 1) % 256

def char_minus():
    """
    perfoms a -
    """
    global memptr
    memcells[memptr] = (memcells[memptr] - 1) % 256

def char_dot():
    """
    performs a .
    """
    global memptr
    sys.stdout.write(chr(memcells[memptr]))

def char_comma():
    """
    performs a ,
    """
    global memptr
    c = ord(sys.stdin.read(1))
    if c != 26:
        memcells[memptr] = c

def loopmap(code):
    """
    maps the "["s to the corresponding "]"s
    """
    opening_brackets_stack = []
    map_loops = {}
    for index,char in enumerate(code):
        if char == "[":
            opening_brackets_stack.append(index)
        elif char == "]":
            try:
                begin = opening_brackets_stack.pop()
                map_loops[begin] = index
            except IndexError:
                raise (ValueError, "Number of ] is more than [")
    if opening_brackets_stack != []:
        raise (ValueError, "Number of [ is more than ]")
    else:
        return map_loops

def evaluate(code):
    """
    evaluates brainfuck code
    """
    global memptr
    map_loops = loopmap(code)
    pc = 0
    stack = []
    while pc < len(code):
        instruction = code[pc]
        if instruction == ">":
            char_gt()
        elif instruction == "<":
            char_lt()
        elif instruction == "+":
            char_plus()
        elif instruction == "-":
            char_minus()
        elif instruction == ".":
            char_dot()
        elif instruction == ",":
            char_comma()
        elif instruction == "[":
            if memcells[memptr] > 0:
                stack.append(pc)
            else:
                pc = map_loops[pc]
        elif instruction == "]":
            pc = stack.pop() - 1
        pc += 1

def filter_code(code):
    """
    Filters the code and ignores anything that's not an instruction
    """
    return list(filter(lambda x: x in ['.', ',', '[', ']', '<', '>', '+', '-'], code))

file = open(sys.argv[1],'r')
cleaned_code = filter_code(file.read())
evaluate(cleaned_code)
file.close()

File: brainfuck.rb
=begin
A simple (probably) brainfuck interpreter
If does not work, blame Jason
=end
require "io/console"
$memcells = Array.new(30000,0)
$memptr = 0
def filter_code(code)
  return code.find_all{ |item| ['>','<','+','-','.',',','[',']'].include? item}
end

def char_lt()
  if $memptr <=0
    abort("Trying to point before starting of memory")
  end
  $memptr-=1
end

def char_gt()
  if $memptr >= 30000 - 1
    abort("Trying to point outside memory")
  end
  $memptr+=1
end

def char_plus()

  $memcells[$memptr] = ($memcells[$memptr] + 1)%256
end

def char_minus()
  $memcells[$memptr] = ($memcells[$memptr] - 1)%256
end

def char_dot()
  STDOUT.print $memcells[$memptr].chr
end

def char_comma()
  c = STDIN.getch.ord
  if c != 26
    $memcells[$memptr] = c
  end
end

def loopmap(code)
  opening_brackets_stack = Array.new
  map_loops = Hash.new

  code.each_with_index do |char,index|
    if char == '['
      opening_brackets_stack.push(index)
    elsif char == ']'
      if opening_brackets_stack.size == 0
        abort("Number of ] more than [")
      end
      start = opening_brackets_stack.pop()
      map_loops[start] = index
    end
  end
  if opening_brackets_stack.size != 0
    abort("Number of [ more than ]")
  else
    return map_loops
  end
end

def evaluate(code)
  pc = 0
  stack = Array.new
  map_loops = loopmap(code)
  while pc < code.length
    instruction = code[pc]
    case instruction
    when '>'
      char_gt()
    when '<'
      char_lt()
    when '+'
      char_plus()
    when '-'
      char_minus()
    when '.'
      char_dot()
    when ','
      char_comma()
    when '['
      if $memcells[$memptr] > 0
        stack.push(pc)
      else
        pc = map_loops[pc]
      end
    when ']'
      pc = stack.pop() - 1
    end
    pc += 1
  end
end

code = Array.new
file = File.open(ARGV[0],"r")
while (m = file.getc)
  code.push(m)
end
cleaned_code = filter_code(code)
evaluate(cleaned_code)
